Zend/zend_alloc.c

```diff
@@ -296,6 +296,20 @@ static bool zend_mm_use_huge_pages = false;
  */
 
 struct _zend_mm_heap {
+    union {                                             /* This union contains security-relevant properties, and is thus made read-only at run-time. */
+        struct {
+            uintptr_t          shadow_key;              /* free slot shadow ptr xor key */
+#if ZEND_MM_CUSTOM
+            struct {
+                void      *(*_malloc)(size_t ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+                void       (*_free)(void*  ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+                void      *(*_realloc)(void*, size_t  ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+            } custom_heap;
+#endif
+        };
+        char padding [ZEND_MM_PAGE_SIZE];
```

Comment:
ZEND_MM_PAGE_SIZE is 4 KB, but that's not necessarily the OS page size (which mprotect uses).

Reply:
The recommended way to get the page size is to call `sysconf(_SC_PAGESIZE)`, but this isn't a compile-time constant (which makes sense, since the page size can be changed). What would be to best™ way to get a reasonable compile-time value?

Reply:
What we could do is to store `custom_heap` and `storage` in a completely separate struct, and address it relatively to the `heap` pointer, e.g.

```
#define ZEND_MM_RO_HEAP(heap) (zend_mm_ro_heap*)((char*)(heap) + REAL_PAGE_SIZE)
```

The extra instructions and memory loads are probably ok as this is only used in slow path.

Reply:
I don't think there is a nice way to force stack-allocated data to be at a specific offset. What kind of structure construction did you have in mind?

Reply:
What I had in mind was to mmap() two pages, use the first one for the heap and the second one for the RO structure.

Reply:
E.g. in `zend_mm_init()`, we could allocate the main chunk and the heap separately:

```
heap = zend_mm_chunk_alloc_int(REAL_PAGE_SIZE * 2, REAL_PAGE_SIZE);
chunk = zend_mm_chunk_alloc_int(ZEND_MM_CHUNK_SIZE, ZEND_MM_CHUNK_SIZE);
chunk->heap = heap;
```